internal/internal_task_pollers.go

```diff
@@ -1011,24 +1036,32 @@ func convertActivityResultToRespondRequestByID(identity, namespace, workflowID,
 			Identity:   identity}
 	}
 
-	var canceledErr *CanceledError
-	if errors.As(err, &canceledErr) {
-		return &workflowservice.RespondActivityTaskCanceledByIdRequest{
-			Namespace:  namespace,
-			WorkflowId: workflowID,
-			RunId:      runID,
-			ActivityId: activityID,
-			Details:    convertErrDetailsToPayloads(canceledErr.details, dataConverter),
-			Identity:   identity}
+	// Only respond with canceled if allowed
+	if cancelAllowed {
+		var canceledErr *CanceledError
+		if errors.As(err, &canceledErr) {
+			return &workflowservice.RespondActivityTaskCanceledByIdRequest{
+				Namespace:  namespace,
+				WorkflowId: workflowID,
+				RunId:      runID,
+				ActivityId: activityID,
+				Details:    convertErrDetailsToPayloads(canceledErr.details, dataConverter),
+				Identity:   identity}
+		}
+		if errors.Is(err, context.Canceled) {
+			return &workflowservice.RespondActivityTaskCanceledByIdRequest{
+				Namespace:  namespace,
+				WorkflowId: workflowID,
+				RunId:      runID,
+				ActivityId: activityID,
+				Identity:   identity}
+		}
 	}
 
-	if errors.Is(err, context.Canceled) {
-		return &workflowservice.RespondActivityTaskCanceledByIdRequest{
-			Namespace:  namespace,
-			WorkflowId: workflowID,
-			RunId:      runID,
-			ActivityId: activityID,
-			Identity:   identity}
+	// If a canceled error is returned but it wasn't allowed, we have to wrap in
+	// an unexpected-cancel application error
+	if _, isCanceledErr := err.(*CanceledError); isCanceledErr {
+		err = fmt.Errorf("unexpected activity cancel error: %w", err)
```

Comment:
When people return CanceledError, they won't expect to see this error. It would be great to give more clear message that is self explanatory so they don't need to seek advise and know exactly what to fix.

Reply:
Suggestion?